File: Prob1.py
from Queue import PriorityQueue
class Twitter(object):
    class Tweet(object):
        def __init__(self,tId,createdAt):
            self.tId = tId
            self.createdAt = createdAt
        
    def __init__(self):
        """
        Initialize your data structure here.
        """
        self.tweets = {}
        self.following = {}
        self.time = 0

    def postTweet(self, userId, tweetId):
        """
        Compose a new tweet.
        :type userId: int
        :type tweetId: int
        :rtype: None
        """
        self.follow(userId,userId)
        if userId not in self.tweets:
            self.tweets[userId] = []
        self.time += 1
        self.tweets[userId].append(self.Tweet(tweetId,self.time))
        

    def getNewsFeed(self, userId):
        """
        Retrieve the 10 most recent tweet ids in the user's news feed. Each item in the news feed must be posted by users who the user followed or by the user herself. Tweets must be ordered from most recent to least recent.
        :type userId: int
        :rtype: List[int]
        """
        q = PriorityQueue()
        followSet = self.following.get(userId,set())
        for user in followSet:
            fTweets = self.tweets.get(user,[])
            for tw in fTweets:
                q.put((tw.createdAt,tw))
                if q.qsize()>10:
                    q.get()
        res = []
        while not q.empty():
            res.append(q.get()[1].tId)
        return res[::-1]
        

    def follow(self, followerId, followeeId):
        """
        Follower follows a followee. If the operation is invalid, it should be a no-op.
        :type followerId: int
        :type followeeId: int
        :rtype: None
        """
        if followerId not in self.following:
            self.following[followerId] = set()
        self.following[followerId].add(followeeId)

    def unfollow(self, followerId, followeeId):
        """
        Follower unfollows a followee. If the operation is invalid, it should be a no-op.
        :type followerId: int
        :type followeeId: int
        :rtype: None
        """
        if followerId in self.following and followerId != followeeId and followeeId in self.following[followerId]:
            self.following[followerId].remove(followeeId)
        


# Your Twitter object will be instantiated and called as such:
# obj = Twitter()
# obj.postTweet(userId,tweetId)
# param_2 = obj.getNewsFeed(userId)
# obj.follow(followerId,followeeId)
# obj.unfollow(followerId,followeeId)

File: prob1.java
// Time Complexity : O(1) 
// Space Complexity : O(n)
// Did this code successfully run on Leetcode : Yes

class Twitter {
    Stack<Pair<Integer,Integer>> tweets;
    HashMap<Integer, Set<Integer>> network;

    public Twitter() {
		tweets = new Stack<>();
		network = new HashMap<>();
    }
    
    public void postTweet(int userId, int tweetId) {
        tweets.push(new Pair(userId, tweetId));
    }
    
    public List<Integer> getNewsFeed(int userId) {
        Stack<Pair<Integer, Integer>> store = new Stack<>();
        List<Integer> ans = new ArrayList<>();
        Set<Integer> following = network.getOrDefault(userId, new HashSet<>());
        
        while(tweets.size()!=0 && ans.size()!=10){
            Pair<Integer, Integer> curr = tweets.pop();
            if((following.contains(curr.getKey()) || curr.getKey()==userId)
               && ans.size()!=10) ans.add(curr.getValue());
            store.push(curr);
        }
        while(store.size()!=0) tweets.push(store.pop());
        return ans;
    }
    
    public void follow(int followerId, int followeeId) {
        Set<Integer> following = network.getOrDefault(followerId, new HashSet<>());
        following.add(followeeId);
        network.put(followerId, following);
    }
    
    public void unfollow(int followerId, int followeeId) {
        Set<Integer> following = network.get(followerId);
        if(following == null) return;
        following.remove(followeeId);
        network.put(followerId, following);
    }
}

